SWIPs/swip-adaptive-node-protocol.md

```diff
@@ -0,0 +1,131 @@
+---
+SWIP: 12
+title: Adaptive nodes capabilities protocol
+author: Louis Holbrook <dev@holbrook.no> (https://holbrook.no), Tim Bansemer (editorial) @FantasticoFox <tim@ethswarm.org>
+status: Draft
+type: Track Networking
+created: 2019-07-11
+---
+
+## Abstract
+
+Adaptive node operation means a node should be able to change the capabilities it has according to changes in available resources.
+
+## Motivation
+
+The peers of the node may need to adjust their behavior towards the node when these capabilities change. Therefore, a protocol is required to inform the peers of changes in capabilities.
+
+## Specification
+
+Upon initial connection, capabilities can be transmitted already within the handshake message.
+
+Capabilities are combinational. Some combination may make more sense than other. Interpreting which combinations are feasible and which are not should be handled by respective nodes, and not be hardwired in the protocol.
+
+For this reason, a bit vector seems the most logical choice. The bit vector can represent the adaptive node operation state within the node, and be embedded into the protocol message as-is.
+Since the capabilities may have to be set from different system modules, the network module should provide an API for registering them. The following structure is proposed:
+
+* the first byte is the identifier of the module (0x00 is bzz, and built in)
+* the second byte is the length `n` of the bit vector containing the flags
+* the last `n` bytes is the bit vector containing the flags
+
+For the individually capability flags we will initially be choosing from, we can see an example of two distinct byte pairs representing tematic categories:
+
+```
+BZZ/SYNC: 			// (registered by default)
+- retrieve		0x00020001
```

Comment:
all these capabilities should be elaborated on right after in terms of what they mean